source/common/router/retry_state_impl.cc

```diff
@@ -27,7 +33,7 @@ RetryStatePtr RetryStateImpl::create(const RetryPolicy& route_policy,
 
   // We short circuit here and do not bother with an allocation if there is no chance we will retry.
   if (request_headers.EnvoyRetryOn() || request_headers.EnvoyRetryGrpcOn() ||
-      route_policy.retryOn()) {
+      route_policy.retryOn() || clusterUseGrid(cluster)) {
```

Comment:
Oh carp, I didn't quite think this out.  If we're going to create a retry policy _and buffer_, I think we're going to want to have explicit configuration.
Before we go any farther here I want Matt's advice.

Hey @mattklein123 can I get API thoughts on this PR?
tl;dr generally users are going to want to send most GET requests over 0-rtt QUIC
Sometimes servers may have 0-rtt handshakes turned off, at which point they send a 425, and the client should retry after the handshake has been confirmed.

I had asked Dan to do this "by default" because I don't want folks having to configure the correct behavior for HTTP/3 but auto-creating a retry policy means buffering bodies and I think we simply shouldn't do that by default.  I think I am Ok auto-creating a retry policy for requests without bodies, but requests with bodies (which is mostly non-idempotent and shouldn't 0-rtt by default) wouldn't retry 425.  
Should I give up on doing the right thing by default and have explicit config or  is it reasonable to auto-retry for bodiless requests, or

Reply:
I started to look through this PR but it's complicated. I will spend some time on this today/tomorrow and response then. Sorry for the delay.

Reply:
@mattklein123 thanks for looking at this PR! To make it easier to understand, this PR, though large, only changes router retry behavior. The majority plumbings are no-op in conn pool. And we are discussing about those seemingly "must have" retries (because they are h3 or 0-RTT) should depends on config knobs or not.